source/extensions/filters/http/cache/cache_filter.cc

```diff
@@ -0,0 +1,220 @@
+         (forwarded_proto->value() == header_values.SchemeValues.Http ||
+          forwarded_proto->value() == header_values.SchemeValues.Https);
+}
+
+bool isCacheableResponse(Http::HeaderMap& headers) {
+  const Http::HeaderEntry* cache_control = headers.CacheControl();
+  // TODO(toddmgreer) fully check for cacheability. See for example
+  // https://github.com/apache/incubator-pagespeed-mod/blob/master/pagespeed/kernel/http/caching_headers.h.
+  if (cache_control) {
+    return !StringUtil::caseFindToken(cache_control->value().getStringView(), ",",
+                                      Http::Headers::get().CacheControlValues.Private);
+  }
+  return false;
+}
+
+HttpCache& getCache(const envoy::config::filter::http::cache::v3alpha::Cache& config) {
+  HttpCacheFactory* factory =
+      Registry::FactoryRegistry<HttpCacheFactory>::getFactory(config.name());
+  if (!factory) {
+    throw ProtoValidationException(
+        fmt::format("Didn't find a registered HttpCacheFactory for '{}'", config.name()), config);
+  }
+  return factory->getCache();
+}
+} // namespace
+
+CacheFilter::CacheFilter(const envoy::config::filter::http::cache::v3alpha::Cache& config,
+                         const std::string&, Stats::Scope&, TimeSource& time_source)
+    : time_source_(time_source), cache_(getCache(config)) {}
+
+void CacheFilter::onDestroy() {
+  lookup_ = nullptr;
+  insert_ = nullptr;
+}
+
+Http::FilterHeadersStatus CacheFilter::decodeHeaders(Http::HeaderMap& headers, bool) {
+  ENVOY_STREAM_LOG(debug, "CacheFilter::decodeHeaders: {}", *decoder_callbacks_, headers);
+  if (!isCacheableRequest(headers)) {
+    ENVOY_STREAM_LOG(debug, "CacheFilter::decodeHeaders ignoring uncacheable request: {}",
+                     *decoder_callbacks_, headers);
+    return Http::FilterHeadersStatus::Continue;
+  }
+  ASSERT(decoder_callbacks_);
+  lookup_ = cache_.makeLookupContext(LookupRequest(headers, time_source_.systemTime()));
+  ASSERT(lookup_);
+
+  CacheFilterSharedPtr self = shared_from_this();
+  ENVOY_STREAM_LOG(debug, "CacheFilter::decodeHeaders starting lookup", *decoder_callbacks_);
+  lookup_->getHeaders([self](LookupResult&& result) { onHeadersAsync(self, std::move(result)); });
+  return Http::FilterHeadersStatus::StopIteration;
+}
+
+Http::FilterHeadersStatus CacheFilter::encodeHeaders(Http::HeaderMap& headers, bool end_stream) {
+  if (lookup_ && isCacheableResponse(headers)) {
+    ENVOY_STREAM_LOG(debug, "CacheFilter::encodeHeaders inserting headers", *encoder_callbacks_);
+    insert_ = cache_.makeInsertContext(std::move(lookup_));
+    insert_->insertHeaders(headers, end_stream);
+  }
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus CacheFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  if (insert_) {
+    ENVOY_STREAM_LOG(debug, "CacheFilter::encodeHeaders inserting body", *encoder_callbacks_);
+    // TODO(toddmgreer) Wait for the cache if necessary.
+    insert_->insertBody(
```

Comment:
@toddmgreer Concerning the buffer size here, is there a presumable value for the huge bodies (i.e. the example with the size of 5 MB on the doc)? Also, could it be configurable by cache provider like "at most X bytes are accepted at a time for body insertions" ?